File: Algorithm/303. Range Sum Query - Immutable.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-08-15 22:29:39
# @Last modified by:   WuLC
# @Last Modified time: 2016-08-15 22:29:46
# @Email: liangchaowu5@gmail.com

class NumArray(object):
    def __init__(self, nums):
        """
        initialize your data structure here.
        :type nums: List[int]
        """
        for i in xrange(1, len(nums)):
            nums[i] += nums[i-1]
        self.nums = nums
        

    def sumRange(self, i, j):
        """
        sum of elements nums[i..j], inclusive.
        :type i: int
        :type j: int
        :rtype: int
        """
        if i == 0:
            return self.nums[j]
        else:
            return self.nums[j] - self.nums[i-1]
        


# Your NumArray object will be instantiated and called as such:
# numArray = NumArray(nums)
# numArray.sumRange(0, 1)
# numArray.sumRange(1, 2)

File: Algorithm/Java/303. Range Sum Query - Immutable.java
/*
* @Author: WuLC
* @Date:   2017-06-29 09:35:38
* @Last Modified by:   WuLC
* @Last Modified time: 2017-06-29 09:39:57
* @Email: liangchaowu5@gmail.com
*/


// calculate presum, with or without additional element is OK
// without additional elemnt
public class NumArray 
{
    private int[] preSum;
    
    public NumArray(int[] nums) 
    {
        preSum = new int[nums.length];
        for(int i = 0; i < nums.length; i++)
        {
            if(i == 0) preSum[i] = nums[i];
            else preSum[i] = preSum[i-1] + nums[i];
        }
    }
    
    public int sumRange(int i, int j) 
    {
        if (i == 0) return preSum[j];
        else return preSum[j] - preSum[i-1];
    }
}


// with additional element
public class NumArray 
{
    private int[] preSum;
    
    public NumArray(int[] nums) 
    {
        preSum = new int[nums.length + 1];
        for(int i = 0; i < nums.length; i++)
            preSum[i + 1] = preSum[i] + nums[i];
    }
    
    public int sumRange(int i, int j) 
    {
        return preSum[j+1] - preSum[i];
    }
}
